### PR TITLE
Weighted QAOA

### PR DESCRIPTION
**Pull request by akotil**
Currently, `light_cone_qaoa` disregards the weights in a given graph. This PR adds "weights" data to the coefficients of Paulis as well as to the calculation of ZZ expectation values to support energy calculation of weighted graphs.

For a given graph to be treated as a weighted graph, the edges must have a "`weight`" attribute. For the edges without this attribute, the weight is assumed to be 1.0.

note: the individual node weights are not taken into account, only edge weights.

### CODE REVIEW

File: large_scale_qaoa/qaoa.py [hunk 1]
```diff
@@ -48,11 +48,11 @@ class ErrorMitigationQAOA:
...
-        self,
-        shots: int,
-        local_correlators: List[Tuple[str, float]],
-        backend,
-        path: Optional[List[int]] = None,
+            self,
+            shots: int,
+            local_correlators: List[Tuple[str, float]],
+            backend,
+            path: Optional[List[int]] = None,
```
Comment (eggerdj) [inline large_scale_qaoa/qaoa.py:55]: I don't think this change is needed, looks like only indentation. Can you revert this?

File: large_scale_qaoa/qaoa.py [hunk 2]
```diff
@@ -116,13 +116,13 @@ def replace_rz_with_barrier(qc: QuantumCircuit):
...
-        transpile_circ: bool = False,
-        remove_rz: bool = False,
-        apply_swaps: bool = True,
+            self,
+            theta,
+            superposition=True,
+            random_cut=None,
+            transpile_circ: bool = False,
+            remove_rz: bool = False,
+            apply_swaps: bool = True,
```
Comment (eggerdj) [inline large_scale_qaoa/qaoa.py:125]: Same comment on indentation.

File: tests/test_graph_construction.py [hunk 3]
```diff
@@ -29,3 +30,16 @@ def test_round_trip(self):
...
+        """Test that we can easily round-trip weighted Pauli the graphs."""
+
+        for seed in range(5):
+            graph1 = nx.random_regular_graph(3, 10, seed=seed)
+            for (u, v) in graph1.edges():
+                graph1.edges[u, v]['weight'] = np.random.normal()
+
+            graph2 = build_graph(build_paulis(graph1))
+
+            self.assertTrue(self._test_edge_equality(graph1, graph2))
```
Comment (eggerdj) [inline tests/test_graph_construction.py:45]: For this test to really make sense you also need to test the `weight` of the graphs in `_test_edge_equality`. I think you will need to modify `_test_edge_equality` and get the edges with `data=True` and then also check that if the weights are present then they should be identical.

**Review by eggerdj — COMMENTED** [review]
Thanks Ayse, this looks almost good to go. Can (i) go over the suggestions and (ii) leave the notebooks (NB) out of the PR? (I.e. do not modify them.) It seems like you did not change the NBs expect from removing the output which makes it harder to refer to them over GitHub.

File: tests/test_light_cone_qaoa.py [hunk 4]
```diff
@@ -1,54 +1,72 @@
...
 from large_scale_qaoa.light_cone_qaoa import LightConeQAOA
 from large_scale_qaoa.graph_utils import build_paulis
 from large_scale_qaoa.qaoa import ErrorMitigationQAOA
 
 
-class TestLightConeQAOA(TestCase):
+class TestLightConeQAOA:
 
-    def setUp(self) -> None:
+    @pytest.fixture()
```
[the remark this answers is not in the document]
Comment (akotil) [reply tests/test_light_cone_qaoa.py:16]: This defines objects that can be commonly used by test cases. For instance, if you define a fixture to return an object, you can pass an argument to a test case with the same name, and the object gets automatically 'injected' to the test case (see the test case `test_sub_correlators` ). It is similar to what the test file had before with a '`setUp`'. `setUp` defined a class attribute to be used across test cases, much like `pytest`'s fixture.

File: tests/test_graph_construction.py (hunk 3 above)
Comment (akotil) [reply tests/test_graph_construction.py:49]: Sorry for overseeing the helper method. Fixed it now.

**Comment by akotil** [comment]
> Thanks Ayse, this looks almost good to go. Can (i) go over the suggestions and (ii) leave the notebooks (NB) out of the PR? (I.e. do not modify them.) It seems like you did not change the NBs expect from removing the output which makes it harder to refer to them over GitHub.

Thanks a lot for the review! I went over the feedback and applied changes.

File: tests/test_light_cone_qaoa.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eggerdj) [reply tests/test_light_cone_qaoa.py:16]: thanks.

File: tests/test_graph_construction.py (hunk 3 above)
Comment (eggerdj) [reply tests/test_graph_construction.py:49]: Thanks

**Review by eggerdj — APPROVED** [review]
LGTM thanks!

**Review by eggerdj — APPROVED** [review]
LGTM thanks!

**Pull request merged**